### PR TITLE
Issue 19646 write directly into tar gzip

### LINKED ISSUE #19646: Push Publish performance
Need to rely on Glowroot (or other tools) to determine the bottlenecks and then work on the strategy to handle those.

### PR DESCRIPTION
**Pull request by freddyucv**
When a bundle is generate to download it is compress in a tar.gzip file.
Before this change we were creating the bundle in a temp directory to after that generate the tar.gzip file. 
Now with this change we are writting directly into the tar.gzip file to improve performan

### CODE REVIEW

File: dotCMS/dependencies.gradle [hunk 1]
```diff
@@ -3,7 +3,7 @@ dependencies {
 
     def eeType = dotcmsReleaseVersion.equals("master") ? '-SNAPSHOT' : ''
 
-    compile group: 'com.dotcms.enterprise', name: 'ee', version: dotcmsReleaseVersion + eeType, changing: true
+    compile group: 'com.dotcms.enterprise', name: 'ee', version: 'issue-19646-write-directly-into-tar-gzip-SNAPSHOT', changing: true
```
Comment (jdotcms) [inline dotCMS/dependencies.gradle:6]: is this ok?

Comment (freddyucv) [reply dotCMS/dependencies.gradle:6]: No I have to remove this after the test run

File: dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java [hunk 2]
```diff
@@ -1412,13 +1412,13 @@ private void addSet(final String assetKey, final AssetTypes assetTypes) {
 		 * The current thread wait until all the dependencies are processed
 		 * @throws ExecutionException
 		 */
-		private void waitForAll() throws ExecutionException {
+		private void startAndWait() throws ExecutionException {
 			final String submitterName = "DependencyManagerSubmitter" + Thread.currentThread().getName();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java:1416]: why are you creating a new submitter per thread

File: dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java [hunk 3]
```diff
@@ -1412,13 +1412,13 @@ private void addSet(final String assetKey, final AssetTypes assetTypes) {
...
 			submitter = DotConcurrentFactory.getInstance().getSubmitter(submitterName,
 					new DotConcurrentFactory.SubmitterConfigBuilder()
-							.poolSize(Config.getIntProperty("MIN_NUMBER_THREAD_TO_EXECUTE_BUNDLER", 1))
+							.poolSize(Config.getIntProperty("MIN_NUMBER_THREAD_TO_EXECUTE_BUNDLER", 10))
 							.maxPoolSize(Config.getIntProperty("MAX_NUMBER_THREAD_TO_EXECUTE_BUNDLER", 40))
-							.queueCapacity(Config.getIntProperty("QUEUE_CAPACITY_TO_EXECUTE_BUNDLER", 500))
+							.queueCapacity(Config.getIntProperty("QUEUE_CAPACITY_TO_EXECUTE_BUNDLER", Integer.MAX_VALUE))
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java:1421]: why max value?

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java [hunk 4]
```diff
@@ -168,6 +165,15 @@ public static void objectToXML ( Object obj, File f, boolean removeFirst ) {
         }
     }
 
+    public static void objectToXML(final Object obj, final OutputStream outputStream) {
+        if (xmlSerializer == null) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java:169]: this should be encapsulated on a XMLSerializerUtil or similar such as
```
XMLSerializerUtil.getInstance().marshal(obj, xmlWriter);
```

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java [hunk 5]
```diff
@@ -211,6 +216,18 @@ public static void objectToJSON( Object obj, File f, boolean removeFirst ) {
     }
 
 
+    public static void objectToJSON( final Object obj, final OutputStream outputStream) {
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java:220]: We already have a class that encapsulates the object mapper

DotObjectMapperProvider.getInstance().getDefaultObjectMapper()

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java [hunk 6]
```diff
@@ -211,6 +216,18 @@ public static void objectToJSON( Object obj, File f, boolean removeFirst ) {
...
+
+        try {
+            mapper.writeValue(outputStream, obj);
+        } catch ( FileNotFoundException e ) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java:224]: it seems you can use 
```
catch (FileNotFoundException | IOException) {
     Logger.error( PublisherUtil.class, e.getMessage(), e );
}
```

File: dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import com.liferay.util.FileUtil;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+
+public abstract class BundleOutput implements Closeable {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java:18]: comment it

File: dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    protected PublisherConfig publisherConfig;
+
+    public BundleOutput(final PublisherConfig publisherConfig){
+        this.publisherConfig = publisherConfig;
+    }
+
+    public abstract OutputStream addFile(String filePath) throws IOException;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java:25]: comment me

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java:25]: comment all methods

File: dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.util.Collection;
+
+public class DirectoryBundleOutput extends BundleOutput {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java:16]: doc it

File: dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+    @VisibleForTesting
+    public DirectoryBundleOutput(final PublisherConfig publisherConfig, final File directoryPath) {
+        super(publisherConfig);
+        directoryRootPath = directoryPath;
+    }
+
+    @Override
+    public boolean useHardLink() {
+        return true;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java:34]: shouldn't be this configurable?

File: dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        fileAbsolute.getParentFile().mkdirs();
+
+        if (!fileAbsolute.exists()) {
+            fileAbsolute.createNewFile();
+        }
+
+        return Files.newOutputStream( fileAbsolute.toPath());
+    }
+
+    @NotNull
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java:65]: is the NotNull needed?

File: dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+
+import javax.ws.rs.NotSupportedException;
+import java.io.*;
+import java.nio.file.Files;
+import java.util.Collection;
+import java.util.zip.GZIPOutputStream;
+
+public class TarGzipBundleOutput extends BundleOutput {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java:16]: Doc

File: dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java:1416]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-abb9df68dc16a9532f21f0164ea5c958a1f3462cce8aa09bcce2dd984da40910R1444

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java (hunk 4 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-827653c0f7910bf99c00088d8a5123e24ca2bf4288dff44aa8b54e067f1535b5R171

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java:217]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-827653c0f7910bf99c00088d8a5123e24ca2bf4288dff44aa8b54e067f1535b5R195

File: dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java (hunk 6 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-827653c0f7910bf99c00088d8a5123e24ca2bf4288dff44aa8b54e067f1535b5L224

File: dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java (hunk 7 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java:16]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-e65f8fbb893f9aa226ba9b0fff84476edb7ab06338dc40aaa0b92c4a49b306eeR12

File: dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java (hunk 8 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java:30]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-e65f8fbb893f9aa226ba9b0fff84476edb7ab06338dc40aaa0b92c4a49b306eeR32

File: dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java (hunk 9 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java:19]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-f2555462ad9b90e9c92d2e27a63788c8a8b50bd6242099982cf202919481710dR16

File: dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java (hunk 10 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/DirectoryBundleOutput.java:37]: really it is 

https://github.com/dotCMS/core/blob/e491040a5cc382fa8a25457ba027679b4afa86b7/dotCMS/src/main/java/com/dotcms/publishing/output/BundleOutput.java#L52

File: dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java (hunk 12 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java:20]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-ad30c1685bed7089a0808e5fb617e74e6f11f5e2c73ed0a7fb395293997caa8bR17

File: dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    private File tarGzipFile;
+    private TarArchiveOutputStream tarArchiveOutputStream;
+    private int GZIP_OUTPUT_STREAM_BUFFER_SIZE = 65536;
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/publishing/output/TarGzipBundleOutput.java]: done https://github.com/dotCMS/core/pull/20173/commits/e491040a5cc382fa8a25457ba027679b4afa86b7#diff-ad30c1685bed7089a0808e5fb617e74e6f11f5e2c73ed0a7fb395293997caa8bR23

**Review by dsilvam — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

**Pull request merged**